mnemonic/src/lib.rs

```diff
@@ -58,6 +63,106 @@ impl Entropy {
     }
 }
 
+impl SecureGenerate for Entropy {
+    fn generate<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let mut bytes = [0u8; Entropy::SIZE];
+        rng.fill_bytes(&mut bytes[..]);
+        bytes.into()
+    }
+}
+
+impl Entropy {
+    const PURPOSE_ID: u32 = 0x42000002;
+    const ENCRYPTION_SALT_SIZE: usize = 16;
+    const ENCRYPTION_KDF_ROUNDS: u32 = 256;
+    const ENCRYPTION_CHECKSUM_SIZE_V3: usize = 2;
+
+    pub fn export_encrypted(&self, key: &[u8]) -> Result<Vec<u8>, String> {
+        let mut salt = [0u8; Entropy::ENCRYPTION_SALT_SIZE];
+        rand_core::OsRng.fill_bytes(&mut salt);
+
+        let mut data = Vec::with_capacity(/*purposeId*/ 4 + Entropy::SIZE);
+        data.extend_from_slice(&Entropy::PURPOSE_ID.to_be_bytes());
+        data.extend_from_slice(self.as_bytes());
+
+        let checksum: [u8; Entropy::ENCRYPTION_CHECKSUM_SIZE_V3] = Blake2bHasher::default()
+            .digest(&data)
+            .as_bytes()[0..Entropy::ENCRYPTION_CHECKSUM_SIZE_V3]
+            .try_into()
+            .unwrap();
+        let mut plaintext = Vec::with_capacity(checksum.len() + data.len());
+        plaintext.extend_from_slice(&checksum);
+        plaintext.extend_from_slice(&data);
+        let ciphertext = otp(
+            &plaintext,
+            key,
+            Entropy::ENCRYPTION_KDF_ROUNDS,
+            &salt,
+            Algorithm::Argon2d,
+        )
+        .map_err(|err| format!("{:?}", err))?;
+
+        let mut buf = Vec::with_capacity(
+            /*version*/ 1 + /*kdf rounds*/ 1 + salt.len() + ciphertext.len(),
+        );
+        buf.extend_from_slice(&3u8.to_be_bytes()); // version
+        buf.extend_from_slice(
+            &((Entropy::ENCRYPTION_KDF_ROUNDS as f32).log2() as u8).to_be_bytes(),
+        );
+        buf.extend_from_slice(&salt);
+        buf.extend_from_slice(&ciphertext);
+
+        Ok(buf)
+    }
+
+    pub fn from_encrypted(buf: &[u8], key: &[u8]) -> Result<Entropy, String> {
+        let version = buf[0];
+        let rounds_log = buf[1];
+        if rounds_log > 32 {
+            return Err(format!("Rounds out-of-bounds: 2^{}", rounds_log));
+        }
+        let rounds = 2u32.pow(rounds_log as u32);
+
+        match version {
+            1 => unimplemented!(),
```

Comment:
Are we going to fully discard older versions?

Reply:
The older versions use a KDF OTP function that core-rs-albatross no longer implements.